{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create possible haplotypes for RN and associated colour palettes\n",
    "\n",
    "#4 gene haplotype order and palette\n",
    "firsta = ['S', 's']\n",
    "seconda = ['A', 'a']\n",
    "thirda = ['B', 'b']\n",
    "fourtha = ['R', 'r']\n",
    "hap_4gene_poss =[]\n",
    "\n",
    "for a in firsta:\n",
    "    for b in seconda:\n",
    "        for c in thirda:\n",
    "            for d in fourtha:\n",
    "                hap_4gene_poss.append(a+b+c+d)\n",
    "                \n",
    "order = [0, 4, 3, 10, 2, 9, 8, 14, 1, 7, 6, 12, 5, 13, 11, 15]\n",
    "hap_4gene_order = dict(zip(hap_4gene_poss, order))\n",
    " \n",
    "num_cols = len(hap_4gene_poss)\n",
    "cols = sns.color_palette(\"husl\", num_cols)\n",
    "col_4gene = dict(zip(hap_4gene_poss, cols))\n",
    "hap_4_col = ['RN_RUNX1_g', 'RN_SRSF2', 'RN_TET2a', 'RN_TET2b_g']\n",
    "\n",
    "#3 gene haplotype order and palette\n",
    "firsta = ['S', 's']\n",
    "seconda = ['A', 'a']\n",
    "fourtha = ['R', 'r']\n",
    "hap_3gene_poss =[] \n",
    "\n",
    "for a in firsta:\n",
    "    for b in seconda:\n",
    "        for d in fourtha:\n",
    "            hap_3gene_poss.append(a+b+d)\n",
    "                \n",
    "order = [0, 4, 3, 10, 2, 9, 8, 14, 1, 7, 6, 12, 5, 13, 11, 15]\n",
    "hap_3gene_order = dict(zip(hap_3gene_poss, order))\n",
    "\n",
    "num_cols = len(hap_3gene_poss)\n",
    "cols = sns.color_palette(\"husl\", num_cols)\n",
    "col_3gene = dict(zip(hap_3gene_poss, cols))\n",
    "hap_3_col = [ 'RN_RUNX1_g', 'RN_SRSF2', 'RN_TET2a']\n",
    "\n",
    "#2 gene haplotype order and palette\n",
    "firsta = ['C', 'c']\n",
    "seconda = ['G', 'g']\n",
    "hap_2gene_poss =[]\n",
    "\n",
    "for a in firsta:\n",
    "    for b in seconda:\n",
    "        hap_2gene_poss.append(a+b)\n",
    "                \n",
    "order = [0, 1, 2, 3]\n",
    "hap_2gene_order = dict(zip(hap_2gene_poss, order))\n",
    "            \n",
    "num_cols = len(hap_2gene_poss)\n",
    "cols = sns.color_palette(\"husl\", num_cols)\n",
    "col_2gene = dict(zip(hap_2gene_poss, cols))\n",
    "hap_2_col = ['RN_RUNX1_c', 'RN_RUNX1_g']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the data and flip into a multi index \n",
    "\n",
    "def data_retrieval(sourcefile, pt_initials):\n",
    "    df = pd.read_csv(sourcefile, header = [0,1,2], index_col = 0, sep='\\t')\n",
    "    df = df.stack([0,1,2])\n",
    "    df = df.reorder_levels([1,0,2,3])\n",
    "    df = df.to_frame()  #puts everything back in a dataframe\n",
    "    df.columns = ['Reads']\n",
    "    df['Plate'] = df.index.get_level_values(0)  #These lines send indexes to columns\n",
    "    df['Well'] = df.index.get_level_values(1)\n",
    "    df['Amplicon'] = df.index.get_level_values(2)\n",
    "    df['Genotype'] = df.index.get_level_values(3)\n",
    "    df[['Patient', 'one', 'two']] = df['Amplicon'].str.split('_', expand = True)\n",
    "    df = df.drop(columns = ['one', 'two'])\n",
    "\n",
    "    #Import information about plate cell type and patient\n",
    "    key = pd.read_excel('../Data/Amp_data/Amplicon_metadata_fixed.xlsx', sheet_name = 'PlateID') #should this be an input? also in next fucntion\n",
    "    key = key.drop(['Cell Origin', 'Plate Nr', 'Plate Name','Nr of cells', 'fcs-fle' ], axis=1)\n",
    "    key.rename(columns = {'Comments2':'Plate'}, inplace = True)\n",
    "    key.rename(columns = {'Cell-group':'Celltype'}, inplace = True)\n",
    "    \n",
    "    #Make a dictionary to associate plates with patients and plate with cell type\n",
    "    plate_pt_dict = dict(zip(key.Plate, key.Patient))\n",
    "    plate_cell_dict = dict(zip(key.Plate, key.Celltype))\n",
    "\n",
    "    #Now just look at data from selected patient, and apply filters to identify cells with enough reads/amplicon\n",
    "    #RN_allele_plate is the key dataset going forward\n",
    "    pt_allele_plate = df.loc[df['Patient'].isin([pt_initials])] #Make df with just RN data\n",
    "    pt_allele_plate = pt_allele_plate.drop(columns = 'Patient') #Drop the Patient ID column and other unwanted cols\n",
    "    pt_allele_plate['Cell_type'] = pt_allele_plate['Plate'].replace(plate_cell_dict)\n",
    "    pt_allele_plate['Plate_Well'] = pt_allele_plate['Plate'].astype(str) + '_' + pt_allele_plate['Well'].astype(str)\n",
    "    #pt_allele_plate = pt_allele_plate.groupby(['Plate', 'Well', 'Amplicon']).sum().unstack()\n",
    "    \n",
    "    return pt_allele_plate\n",
    "\n",
    "def call_haps (data, haps, reads, cutoff):\n",
    "\n",
    "#data is the df created by data_retrieval\n",
    "#haps is the number of haplotypes\n",
    "#reads is the number of reads per amplicon\n",
    "#cutoff is the threshold for calling an amplicon as mutant\n",
    "    \n",
    "    #Import information about plate cell type and patient\n",
    "    key = pd.read_excel('../Data/Amp_data/Amplicon_metadata_fixed.xlsx', sheet_name = 'PlateID')\n",
    "    key = key.drop(['Cell Origin', 'Plate Nr', 'Plate Name','Nr of cells', 'fcs-fle' ], axis=1)\n",
    "    key.rename(columns = {'Comments2':'Plate'}, inplace = True)\n",
    "    key.rename(columns = {'Cell-group':'Celltype'}, inplace = True)\n",
    "    \n",
    "    #Make a dictionary to associate plates with patients and plate with cell type\n",
    "    plate_pt_dict = dict(zip(key.Plate, key.Patient))\n",
    "    plate_cell_dict = dict(zip(key.Plate, key.Celltype))\n",
    "    \n",
    "    #Create possible haplotypes for RN and associated colour palettes\n",
    "\n",
    "    #4 gene haplotype order and palette\n",
    "    firsta = ['S', 's']\n",
    "    seconda = ['A', 'a']\n",
    "    thirda = ['B', 'b']\n",
    "    fourtha = ['R', 'r']\n",
    "    hap_4gene_poss =[]\n",
    "\n",
    "    for a in firsta:\n",
    "        for b in seconda:\n",
    "            for c in thirda:\n",
    "                for d in fourtha:\n",
    "                    hap_4gene_poss.append(a+b+c+d)\n",
    "\n",
    "    order = [0, 4, 3, 10, 2, 9, 8, 14, 1, 7, 6, 12, 5, 13, 11, 15]\n",
    "    hap_4gene_order = dict(zip(hap_4gene_poss, order))\n",
    "\n",
    "    num_cols = len(hap_4gene_poss)\n",
    "    cols = sns.color_palette(\"husl\", num_cols)\n",
    "    col_4gene = dict(zip(hap_4gene_poss, cols))\n",
    "    hap_4_col = ['RN_RUNX1_g', 'RN_SRSF2', 'RN_TET2a', 'RN_TET2b_g']\n",
    "\n",
    "    #3 gene haplotype order and palette\n",
    "    firsta = ['S', 's']\n",
    "    seconda = ['A', 'a']\n",
    "    fourtha = ['R', 'r']\n",
    "    hap_3gene_poss =[] \n",
    "\n",
    "    for a in firsta:\n",
    "        for b in seconda:\n",
    "            for d in fourtha:\n",
    "                hap_3gene_poss.append(a+b+d)\n",
    "\n",
    "    order = [0, 4, 3, 10, 2, 9, 8, 14, 1, 7, 6, 12, 5, 13, 11, 15]\n",
    "    hap_3gene_order = dict(zip(hap_3gene_poss, order))\n",
    "\n",
    "    num_cols = len(hap_3gene_poss)\n",
    "    cols = sns.color_palette(\"husl\", num_cols)\n",
    "    col_3gene = dict(zip(hap_3gene_poss, cols))\n",
    "    hap_3_col = [ 'RN_RUNX1_g', 'RN_SRSF2', 'RN_TET2a']\n",
    "\n",
    "    #2 gene haplotype order and palette\n",
    "    firsta = ['C', 'c']\n",
    "    seconda = ['G', 'g']\n",
    "    hap_2gene_poss =[]\n",
    "\n",
    "    for a in firsta:\n",
    "        for b in seconda:\n",
    "            hap_2gene_poss.append(a+b)\n",
    "\n",
    "    order = [0, 1, 2, 3]\n",
    "    hap_2gene_order = dict(zip(hap_2gene_poss, order))\n",
    "\n",
    "    num_cols = len(hap_2gene_poss)\n",
    "    cols = sns.color_palette(\"husl\", num_cols)\n",
    "    col_2gene = dict(zip(hap_2gene_poss, cols))\n",
    "    hap_2_col = ['RN_RUNX1_c', 'RN_RUNX1_g']\n",
    "    \n",
    "    if haps == 4:\n",
    "        cols = hap_4_col\n",
    "    \n",
    "    elif haps == 2:\n",
    "        cols = hap_2_col\n",
    "        \n",
    "    elif haps == 3:\n",
    "        cols = hap_3_col\n",
    "        \n",
    "    else:\n",
    "        print('Enter 2, 3, or 4 haplotypes')\n",
    "    \n",
    "    #Group the data and apply filters\n",
    "    df = data.copy()\n",
    "    df = df.groupby(['Plate', 'Well', 'Amplicon']).sum().unstack()\n",
    "    df.columns = ['RN_RUNX1_c','RN_RUNX1_g','RN_SRSF2','RN_TET2a','RN_TET2b_c','RN_TET2b_g']\n",
    "    \n",
    "    df = df.loc[(df[cols] >= reads).all(axis=1)] #df1 contains just the rows with cells we want - use this to create a filter or key\n",
    "    df['Plate'] = df.index.get_level_values(0)  #These lines send indexes to columns\n",
    "    df['Well'] = df.index.get_level_values(1)\n",
    "    df['Plate_Well'] = df['Plate'].astype(str) + '_' + df['Well'].astype(str)\n",
    "    wells = df['Plate_Well'].drop_duplicates().to_list() \n",
    "    print(f'Cells with {reads} reads for {haps} genes = ', len(wells))\n",
    "    \n",
    "    df2 = data.copy()\n",
    "    df2 = df2[df2['Plate_Well'].isin(wells)]\n",
    "    df2 = df2[df2['Amplicon'].isin(cols)]\n",
    "    \n",
    "    #Calculate the allele frequency\n",
    "    df2 = df2.iloc[:, 0:1].unstack(level = 3)\n",
    "    df2['Total'] = df2.iloc[: , 0] + df2.iloc[: , 1]\n",
    "    df2['Mut_freq'] = df2.iloc[:, 0]/df2['Total']\n",
    "\n",
    "    #Assign Wt or MT to each allele\n",
    "    df2 = df2.drop(columns = ['Reads', 'Total'])\n",
    "\n",
    "    conditions = [(df2['Mut_freq'] <= cutoff), (df2['Mut_freq']) > cutoff ]\n",
    "    values = ['w', 'm']\n",
    "    df2['Genotype'] = np.select(conditions, values)\n",
    "    df2 = df2.drop(columns = ['Mut_freq']).unstack(2)\n",
    "    df2.columns = cols\n",
    "    \n",
    "    if 'RN_RUNX1_g' in df2.columns:\n",
    "        df2.loc[:,'RN_RUNX1_g'].replace({'w':'R','m':'r' }, inplace = True)\n",
    "        \n",
    "    if 'RN_SRSF2' in df2.columns:  \n",
    "        df2.loc[:,'RN_SRSF2'].replace({'w':'S','m':'s' }, inplace = True)\n",
    "        \n",
    "    if 'RN_TET2a' in df2.columns:     \n",
    "        df2.loc[:,'RN_TET2a'].replace({'w':'A','m':'a' }, inplace = True)\n",
    "        \n",
    "    if 'RN_TET2b_g' in df2.columns:\n",
    "        df2.loc[:,'RN_TET2b_g'].replace({'w':'B','m':'b' }, inplace = True)\n",
    "        \n",
    "    if 'RN_RUNX1_c' in df2.columns:   \n",
    "        df2.loc[:,'RN_RUNX1_c'].replace({'w':'C','m':'c' }, inplace = True)\n",
    "        \n",
    "    df2['Haplotype'] = 'x'\n",
    "\n",
    "    for idx, row in df2.iterrows():\n",
    "        \n",
    "        if haps == 3:\n",
    "            a = row['RN_SRSF2'] + row['RN_TET2a'] + row['RN_RUNX1_g']\n",
    "        elif haps == 4:\n",
    "            a = row['RN_SRSF2'] + row['RN_TET2a'] + row['RN_TET2b_g'] + row['RN_RUNX1_g']\n",
    "        elif haps ==2:\n",
    "            a = row['RN_RUNX1_c'] + row['RN_RUNX1_g']\n",
    "        \n",
    "        row['Haplotype'] = row['Haplotype'].replace('x', a)   \n",
    "    \n",
    "    df2['Sort_cell_type'] = df2.index.get_level_values(0)\n",
    "    df2['Sort_cell_type'] = df2['Sort_cell_type'].replace(plate_cell_dict)\n",
    "    df2['Plate'] = df2.index.get_level_values(0)\n",
    "    df2['Well'] = df2.index.get_level_values(1)\n",
    "    df2['Plate_Well'] = df2['Plate'].astype(str) + '_' + df2['Well'].astype(str)\n",
    "    df2 = df2.drop(columns = cols)\n",
    "    df2 = df2.drop(columns = ['Plate', 'Well'])\n",
    "\n",
    "    return df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#QC just checking the output is the same.\n",
    "#The columns (L-R) are; [('Reads', 'RN_RUNX1_c'), ('Reads', 'RN_RUNX1_g'), ('Reads', 'RN_SRSF2'), ('Reads', 'RN_TET2a'), ('Reads', 'RN_TET2b_c'), ('Reads', 'RN_TET2b_g')]\n",
    "#Cells with 100 reads for 3 genes (plate1) =  794\n",
    "#Cells with 15 reads for 4 genes (plate2) =  162\n",
    "#Cells with 50 reads for RUNX1 cDNA and gDNA (plate3) =  1645\n",
    "#Cells with 10 reads for 4 genes (plate4) =  253\n",
    "#Cells with 50 reads for 3 genes (plate5) =  1234"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cells with 100 reads for 3 genes =  794\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Haplotype</th>\n",
       "      <th>Sort_cell_type</th>\n",
       "      <th>Plate_Well</th>\n",
       "      <th>FSC-A</th>\n",
       "      <th>FSC-W</th>\n",
       "      <th>FSC-H</th>\n",
       "      <th>SSC-A</th>\n",
       "      <th>SSC-W</th>\n",
       "      <th>SSC-H</th>\n",
       "      <th>CD34-PE</th>\n",
       "      <th>...</th>\n",
       "      <th>CD66b_pos</th>\n",
       "      <th>CD66b_neg</th>\n",
       "      <th>CD16_pos</th>\n",
       "      <th>CD16_neg</th>\n",
       "      <th>CD14_pos</th>\n",
       "      <th>CD14_neg</th>\n",
       "      <th>nBC</th>\n",
       "      <th>NE</th>\n",
       "      <th>CD14pos_mono</th>\n",
       "      <th>CD14neg_mono</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAR</td>\n",
       "      <td>nBCs</td>\n",
       "      <td>AS-194_A12</td>\n",
       "      <td>78240.600</td>\n",
       "      <td>85805.690</td>\n",
       "      <td>59758.0</td>\n",
       "      <td>57048.51</td>\n",
       "      <td>77708.920</td>\n",
       "      <td>48112.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAR</td>\n",
       "      <td>nBCs</td>\n",
       "      <td>AS-194_A13</td>\n",
       "      <td>83133.000</td>\n",
       "      <td>91049.234</td>\n",
       "      <td>59838.0</td>\n",
       "      <td>23547.42</td>\n",
       "      <td>82927.820</td>\n",
       "      <td>18609.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAR</td>\n",
       "      <td>nBCs</td>\n",
       "      <td>AS-194_A16</td>\n",
       "      <td>70706.695</td>\n",
       "      <td>87015.450</td>\n",
       "      <td>53253.0</td>\n",
       "      <td>28536.00</td>\n",
       "      <td>76133.170</td>\n",
       "      <td>24564.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SaR</td>\n",
       "      <td>nBCs</td>\n",
       "      <td>AS-194_A22</td>\n",
       "      <td>50627.700</td>\n",
       "      <td>82859.350</td>\n",
       "      <td>40043.0</td>\n",
       "      <td>17333.88</td>\n",
       "      <td>79512.375</td>\n",
       "      <td>14287.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAR</td>\n",
       "      <td>nBCs</td>\n",
       "      <td>AS-194_B13</td>\n",
       "      <td>70561.800</td>\n",
       "      <td>80219.586</td>\n",
       "      <td>57646.0</td>\n",
       "      <td>13657.26</td>\n",
       "      <td>70497.960</td>\n",
       "      <td>12696.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>SaR</td>\n",
       "      <td>NEs</td>\n",
       "      <td>AS-195_O22</td>\n",
       "      <td>131130.000</td>\n",
       "      <td>99892.310</td>\n",
       "      <td>86030.0</td>\n",
       "      <td>127044.36</td>\n",
       "      <td>90787.920</td>\n",
       "      <td>91708.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>SAR</td>\n",
       "      <td>NEs</td>\n",
       "      <td>AS-195_P18</td>\n",
       "      <td>128855.695</td>\n",
       "      <td>105181.260</td>\n",
       "      <td>80287.0</td>\n",
       "      <td>118534.02</td>\n",
       "      <td>94177.680</td>\n",
       "      <td>82485.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>saR</td>\n",
       "      <td>NEs</td>\n",
       "      <td>AS-189_C2</td>\n",
       "      <td>115577.090</td>\n",
       "      <td>102051.414</td>\n",
       "      <td>74222.0</td>\n",
       "      <td>118898.55</td>\n",
       "      <td>91965.380</td>\n",
       "      <td>84729.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790</th>\n",
       "      <td>saR</td>\n",
       "      <td>NEs</td>\n",
       "      <td>AS-189_D9</td>\n",
       "      <td>148525.200</td>\n",
       "      <td>110377.480</td>\n",
       "      <td>88186.0</td>\n",
       "      <td>141337.60</td>\n",
       "      <td>98506.890</td>\n",
       "      <td>94031.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791</th>\n",
       "      <td>SaR</td>\n",
       "      <td>NEs</td>\n",
       "      <td>AS-189_E2</td>\n",
       "      <td>127727.090</td>\n",
       "      <td>98600.890</td>\n",
       "      <td>84895.0</td>\n",
       "      <td>90046.74</td>\n",
       "      <td>90359.730</td>\n",
       "      <td>65309.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>792 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Haplotype Sort_cell_type  Plate_Well       FSC-A       FSC-W    FSC-H  \\\n",
       "0         SAR           nBCs  AS-194_A12   78240.600   85805.690  59758.0   \n",
       "1         SAR           nBCs  AS-194_A13   83133.000   91049.234  59838.0   \n",
       "2         SAR           nBCs  AS-194_A16   70706.695   87015.450  53253.0   \n",
       "3         SaR           nBCs  AS-194_A22   50627.700   82859.350  40043.0   \n",
       "4         SAR           nBCs  AS-194_B13   70561.800   80219.586  57646.0   \n",
       "..        ...            ...         ...         ...         ...      ...   \n",
       "787       SaR            NEs  AS-195_O22  131130.000   99892.310  86030.0   \n",
       "788       SAR            NEs  AS-195_P18  128855.695  105181.260  80287.0   \n",
       "789       saR            NEs   AS-189_C2  115577.090  102051.414  74222.0   \n",
       "790       saR            NEs   AS-189_D9  148525.200  110377.480  88186.0   \n",
       "791       SaR            NEs   AS-189_E2  127727.090   98600.890  84895.0   \n",
       "\n",
       "         SSC-A      SSC-W    SSC-H  CD34-PE  ...  CD66b_pos  CD66b_neg  \\\n",
       "0     57048.51  77708.920  48112.0      NaN  ...      False       True   \n",
       "1     23547.42  82927.820  18609.0      NaN  ...      False       True   \n",
       "2     28536.00  76133.170  24564.0      NaN  ...      False       True   \n",
       "3     17333.88  79512.375  14287.0      NaN  ...      False       True   \n",
       "4     13657.26  70497.960  12696.0      NaN  ...      False       True   \n",
       "..         ...        ...      ...      ...  ...        ...        ...   \n",
       "787  127044.36  90787.920  91708.0      NaN  ...       True      False   \n",
       "788  118534.02  94177.680  82485.0      NaN  ...       True      False   \n",
       "789  118898.55  91965.380  84729.0      NaN  ...       True      False   \n",
       "790  141337.60  98506.890  94031.0      NaN  ...       True      False   \n",
       "791   90046.74  90359.730  65309.0      NaN  ...       True      False   \n",
       "\n",
       "     CD16_pos  CD16_neg  CD14_pos  CD14_neg    nBC     NE CD14pos_mono  \\\n",
       "0       False      True     False      True   True  False        False   \n",
       "1       False      True     False      True   True  False        False   \n",
       "2       False      True     False      True   True  False        False   \n",
       "3       False      True     False      True   True  False        False   \n",
       "4       False      True     False      True   True  False        False   \n",
       "..        ...       ...       ...       ...    ...    ...          ...   \n",
       "787      True     False     False      True  False   True        False   \n",
       "788      True     False     False      True  False   True        False   \n",
       "789      True     False     False      True  False   True        False   \n",
       "790      True     False     False      True  False   True        False   \n",
       "791      True     False      True     False  False   True        False   \n",
       "\n",
       "    CD14neg_mono  \n",
       "0          False  \n",
       "1          False  \n",
       "2          False  \n",
       "3          False  \n",
       "4          False  \n",
       "..           ...  \n",
       "787        False  \n",
       "788        False  \n",
       "789        False  \n",
       "790        False  \n",
       "791        False  \n",
       "\n",
       "[792 rows x 65 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datafile = '../Data/Amp_data/allele_counts.tsv'\n",
    "index_data = pd.read_csv('../Data/Amp_data/RN_comp_celltype_assignment.tsv', index_col = 0, sep = '\\t')\n",
    "RN_plate = data_retrieval(datafile, 'RN')  \n",
    "RN_haps = call_haps(RN_plate, 3, 100, 0.1) #hap number, read cut, proprtion mutated\n",
    "RN_haps_indexed = pd.merge(RN_haps, index_data, on = 'Plate_Well')\n",
    "RN_haps_indexed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_hap_dist_sort_type(data):\n",
    "    \n",
    "    #4 gene haplotype order and palette\n",
    "    firsta = ['S', 's']\n",
    "    seconda = ['A', 'a']\n",
    "    thirda = ['B', 'b']\n",
    "    fourtha = ['R', 'r']\n",
    "    hap_4gene_poss =[]\n",
    "\n",
    "    for a in firsta:\n",
    "        for b in seconda:\n",
    "            for c in thirda:\n",
    "                for d in fourtha:\n",
    "                    hap_4gene_poss.append(a+b+c+d)\n",
    "\n",
    "    order = [0, 4, 3, 10, 2, 9, 8, 14, 1, 7, 6, 12, 5, 13, 11, 15]\n",
    "    hap_4gene_order = dict(zip(hap_4gene_poss, order))\n",
    "\n",
    "    num_cols = len(hap_4gene_poss)\n",
    "    cols = sns.color_palette(\"husl\", num_cols)\n",
    "    col_4gene = dict(zip(hap_4gene_poss, cols))\n",
    "    hap_4_col = ['RN_RUNX1_g', 'RN_SRSF2', 'RN_TET2a', 'RN_TET2b_g']\n",
    "\n",
    "    #3 gene haplotype order and palette\n",
    "    firsta = ['S', 's']\n",
    "    seconda = ['A', 'a']\n",
    "    fourtha = ['R', 'r']\n",
    "    hap_3gene_poss =[] \n",
    "\n",
    "    for a in firsta:\n",
    "        for b in seconda:\n",
    "            for d in fourtha:\n",
    "                hap_3gene_poss.append(a+b+d)\n",
    "\n",
    "    order = [0, 4, 3, 10, 2, 9, 8, 14, 1, 7, 6, 12, 5, 13, 11, 15]\n",
    "    hap_3gene_order = dict(zip(hap_3gene_poss, order))\n",
    "\n",
    "    num_cols = len(hap_3gene_poss)\n",
    "    cols = sns.color_palette(\"husl\", num_cols)\n",
    "    col_3gene = dict(zip(hap_3gene_poss, cols))\n",
    "    hap_3_col = [ 'RN_RUNX1_g', 'RN_SRSF2', 'RN_TET2a']\n",
    "\n",
    "    #2 gene haplotype order and palette\n",
    "    firsta = ['C', 'c']\n",
    "    seconda = ['G', 'g']\n",
    "    hap_2gene_poss =[]\n",
    "\n",
    "    for a in firsta:\n",
    "        for b in seconda:\n",
    "            hap_2gene_poss.append(a+b)\n",
    "\n",
    "    order = [0, 1, 2, 3]\n",
    "    hap_2gene_order = dict(zip(hap_2gene_poss, order))\n",
    "\n",
    "    num_cols = len(hap_2gene_poss)\n",
    "    cols = sns.color_palette(\"husl\", num_cols)\n",
    "    col_2gene = dict(zip(hap_2gene_poss, cols))\n",
    "    hap_2_col = ['RN_RUNX1_c', 'RN_RUNX1_g']\n",
    "    \n",
    "    #rename the input data and work out how many haplotypes it has\n",
    "    df3 = data.copy()\n",
    "    haps = len(df3.iloc[0,0])\n",
    "    print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c\n"
     ]
    }
   ],
   "source": [
    "plot_hap_dist_sort_type(RN_haps)\n",
    "\n",
    "#Ready to add in plotting info, need to work out whihc plot to use (depending on )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-12-c51401a78f3d>:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  index_data = pd.read_csv('../Data/Amp_data/RN_comp_celltype_assignment.tsv', sep = '/t')\n"
     ]
    }
   ],
   "source": [
    "index_data = pd.read_csv('../Data/Amp_data/RN_comp_celltype_assignment.tsv', sep = '/t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
